File: dataset_parser/coders/ca/decoder_ca.py
import sys
sys.path.append('../')

from coders.cols.decoder_cols import DecoderCols
from coders.ca.window_ca import Point, Line


class DecoderCA(DecoderCols):
    def __init__(self, input_path, input_filename, output_csv, params):
        super(DecoderCA, self).__init__(input_path, input_filename, output_csv)
        self.window_size_bit_length = params['max_window_size'].bit_length()

    def _decode_column(self):
        self.row_index, self.column = 0, []
        previous_value = None

        while self.row_index < self.data_rows_count:
            window_length, value = self._decode_window()
            if value == 'N' or previous_value == value or window_length == 1:
                window = [value] * window_length
            else:
                window = self._create_window(previous_value, value, window_length)
            previous_value = value
            self.column.extend(window)
            self.row_index += window_length

        return self.column

    def _decode_window(self, _=None):
        window_length = self.input_file.read_int(self.window_size_bit_length)
        value = self._decode_value_raw(self.row_index, self.column_index)
        return [window_length, value]

    #
    # PRE: window_length >= 2
    #
    @classmethod
    def _create_window(cls, previous_value, value, window_length):
        first_point = Point(0, previous_value)
        last_point = Point(window_length, value)
        line = Line(first_point, last_point)

        window = []
        for x in range(window_length):
            point = Point(x + 1, 0)  # y doesn't matter
            y = line.y_intersection(point)
            window.append(int(round(y)))

        if window[-1] != value:
            # print "first point:", 0, previous_value
            # print "last point:", window_length, value
            # print window
            # point = Point(7, 0)
            # print "line.point.x", line.point.x
            # print "line.point.y", line.point.y
            # print "line.m", line.m
            # y = line.y_intersection(point)
            # print "y", y
            # print "int(round(y))", int(round(y))
            raise StandardError("These two values must match. window[-1]=%s, value=%s" % (window[-1], value))

        return window

File: cpp_project/src/coders/ca/decoder_ca.cpp

#include "decoder_ca.h"

#include "string_utils.h"
#include "ca_line.h"
#include <cmath>

void DecoderCA::setCoderParams(int max_window_size_){
    max_window_size_bit_length = StringUtils::bitLength(max_window_size_);
}


std::vector<std::string> DecoderCA::decodeColumn(){
    std::vector<std::string> column;
    std::string previous_value = "random"; // any value that cannot be read from the csv
    row_index = 0;

    while (row_index < data_rows_count){
        std::cout << "row_index = " << row_index << std::endl;
        decodeWindow(column);

        if (current_value == Constants::NO_DATA || previous_value == current_value || current_window_size == 1){
            for (int i=0; i < current_window_size; i++){
                column.push_back(current_value);
                row_index++;
            }
        }
        else {
            createWindow(column, previous_value);
        }
        previous_value = current_value;
    }
    return column;
}

void DecoderCA::decodeWindow(std::vector<std::string> & column){
    current_window_size = input_file.getInt(max_window_size_bit_length);
    current_value = decodeValueRaw();
}

void DecoderCA::createWindow(std::vector<std::string> & column, std::string previous_value){
    int previous_value_int = std::stoi(previous_value);
    int current_value_int = std::stoi(current_value);

    CAPoint first_point = CAPoint(0, previous_value_int);
    std::cout << "first_point = (0," << previous_value_int << ")" <<  std::endl;
    int last_point_time_delta = 0;
    for (int i=0; i < current_window_size; i++){ last_point_time_delta += time_delta_vector.at(row_index + i); }
    CAPoint last_point = CAPoint(last_point_time_delta, current_value_int);
    std::cout << "last_point = (" << last_point_time_delta << "," << current_value_int << ")" <<  std::endl;
    CALine line = CALine(first_point, last_point);

    int current_sum = 0;
    int time_delta = 0;

    for (int i=0; i < current_window_size; i++){
//        if (i > 0) { time_delta = time_delta_vector.at(row_index); }
        time_delta = time_delta_vector.at(row_index);
        current_sum += time_delta;

        std::cout << "current_sum = " << current_sum <<  std::endl;
        CAPoint point = CAPoint(current_sum, 0); // y doesn't matter
        double y = line.yIntersection(point);
        int val = std::round(y);
        std::string val_str = std::to_string(val);
        std::cout << "val_str = " << val_str <<  std::endl;
        column.push_back(val_str);
        row_index++;
        if (i == current_window_size - 1){
            assert(val == current_value_int);
        }
    }
}
